src/script/interpreter.cpp

```diff
@@ -1981,9 +1981,16 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
                 return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
             }
 
-            const bool op_cat_is_op_success = !(flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT);
             // New opcodes will be listed here. May use a different sigversion to modify existing opcodes.
-            if (IsOpSuccess(opcode) || (op_cat_is_op_success && opcode == OP_CAT)) {
+            if (IsOpSuccess(opcode)) {
+                if (opcode == OP_CAT) {
+                    if (flags & SCRIPT_VERIFY_DISCOURAGE_TAPSCRIPT_OP_CAT) {
+                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_CAT);
+                    } else if (!(flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT)) {
+                        // op_success behaviour
+                        return set_success(serror);
+                    }
+                }
```

Comment:
If `DISCOURAGE_OP_CAT` is not set, then this will either opt-in to OP_SUCCESS behaviour because `TAPSCRIPT_OP_CAT` is not set, or continue on, and hit OP_SUCCESS behaviour anyway. Should have an `else {`.

Reply:
Yes, I think this would be easier to parse like this:

```c++
               // New opcodes will be listed here. May use a different sigversion to modify existing opcodes.
                if (opcode == OP_CAT && (flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT)) {
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_TAPSCRIPT_OP_CAT) {
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_CAT);
                    } 
                } else if (IsOpSuccess(opcode)) {
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
                    }
                    return set_success(serror);
                }
```

Makes it clear that the behavior is unchanged if OP_CAT is not active.

Reply:
Additionally @halseth proposed change has the advantage that if CAT is removed from `IsOpSuccess` all the flag checks are not skipped.

Reply:
I think this all depends on when you want the `discourage` flag to have an effect. I was under the impression that the discourage-cat flag only had any effect after CAT was active (since before activation it was already OP_SUCCESS and hence discouraged), but maybe this is wrong?

Reply:
And if that's the case would we even need a new discourage flag? We can rely on OP_SUCCESS already being discouraged before activation of OP_CAT.

Reply:
Thanks for the added context, very helpful!

> The SCRIPT_VERIFY_* flags try to each be a soft-fork: that is, if you accept a tx when SCRIPT_VERIFY_X is set, you should also accept it when SCRIPT_VERIFY_X is removed

Ok, that's a useful definition. With that in mind I think your outlined steps make sense. It would look like this:

```c++
if (opcode == OP_CAT) {
    // If discouraged, error whether CAT is active or not.
    if (flags & SCRIPT_VERIFY_DISCOURAGE_TAPSCRIPT_OP_CAT) {
        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_CAT);
    }

    // If not activated, fall back to success behavior.
    if !(flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT) {
        return set_success(serror);
    }
} else if (IsOpSuccess(opcode)) {
    if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
    }
    return set_success(serror);
}
...

bool IsOpSuccess(opcode) {
        if opcode == OP_CAT {
                // Doesn't really matter what we return, can
                // remove the OP_CAT clause.
                return false
        }
        ...
}
```

Removing `DISCOURAGE_CAT` would remove an error case, conforming to your definition.

This has a caveat though: the behavior would be equivalent to just setting `VERIFY_OP_CAT` on activation date, simplifying the code somewhat:

```c++
if (opcode == OP_CAT && (flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT)) {
        // After activation we'll start verifying the opcode.
} else if (IsOpSuccess(opcode)) {
    if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
    }
    return set_success(serror);
}
...

bool IsOpSuccess(opcode) {
        if (opcode == OP_CAT) {
                return true
        }
        ...
}
```

This does however not conform to the definition anymore, since removing `VERIFY_OP_CAT` could result in discouraged op_success handling, taking a transaction from accepted to rejected (this is only true for mempool handling of course, since discourage flags are not consensus enforced).

Functionally I think it's equivalent though, as you would not really remove `VERIFY_OP_CAT` after activation, as that would be a hard fork.  Code wise I would prefer the simpler version since there is one less flag to test with other combinations, and it would be easier to generalize for future soft-forks.

Reply:
Maybe to make it even clearer that OP_CAT retains success behavior before activation: 

```c++
if (IsOpSuccess(opcode)) {
   if (opcode == OP_CAT && (flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT)) {
        // After activation we'll start verifying the opcode.
    } else if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
    } else {
        return set_success(serror);
    }
}
...

bool IsOpSuccess(opcode) {
        if (opcode == OP_CAT) {
                return true
        }
        ...
}
```

Reply:
I tend to agree with the last couple comments made. It seems like `DISCOURAGE_CAT` is just being used as a temporary flag to decouple OP_SUCCESS activation from OP_CAT activation. But the same can be achieved with something like 
```cpp
if (opcode == OP_CAT && (flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT)) {
        // After activation we'll start verifying the opcode.
} else if (IsOpSuccess(opcode)) {
    if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
        return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
    }
    return set_success(serror);
}
```
And leaving `is_opsuccess` as is. I think this adheres to the definition.
OP_CAT will be executed to the new rules even if OP_SUCCESS is discouraged.
and vice-versa. 

 @ajtowns  Lmk if you agree.

Reply:
> Functionally I think it's equivalent though, as you would not really remove `VERIFY_OP_CAT` after activation, as that would be a hard fork.

That's mixing up two different things: `VERIFY_*` flags operate on individual transactions and include both policy and consensus rules; we design each flag to add more restrictions to make it easier to reason about the behaviour so that we can avoid introducing bugs. Activation occurs in the context of a chain, and changes the selection of `VERIFY_*` flags that we apply when validating blocks.

> another benefit of the simpler version is that you avoid having to define a flag bit to mean something when you only need it temporarily.

When the flag is no longer needed, we can delete it from the codebase. That's fine. But we won't know it's no longer needed until the soft fork is activated on mainnet and buried, and that is likely longer than the support lifetime for the inquisition version the code's introduced in.

>  It seems like DISCOURAGE_CAT is just being used as a temporary flag to decouple OP_SUCCESS activation from OP_CAT activation. But the same can be achieved with something like

No, it can't.

Consider the following three scripts: (1) `OP_CAT OP_RETURN`; (2) `0 1 OP_CAT`; (3) `OP_SUBSTR OP_RETURN`. Those should behave as follows:

| CAT Activated? | Mempool/Block | CAT+RET | CAT | SUB+RET |
| --- | ----- | ---- | ---- | ---- |
| N | Block | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Y | Block | :x: | :heavy_check_mark: | :heavy_check_mark: |
| | | | | |
| N | Mempool | :x: | :x: | :x: |
| Y | Mempool | :x: | :heavy_check_mark: | :x: |

(Note that OP_SUCCESS behaviour means the script just succeeds immediately wherever the OP_SUCCESSx appears in the script, so the OP_RETURNs are never seen)

The case of Block/CAT+RET going from yes to no when activation occurs is handled by adding `VERIFY_OP_CAT`; the case of Mempool/CAT going from no to yes when activation occurs is handled by removing `VERIFY_DISCOURAGE_OP_CAT` from the additional mempool validation rules.

If you wanted to have just two flags rather than three while still having each flag only add restrictions, you'd need:

 * no flags: everything works
 * verify_cat: cat+ret fails, everything else works
 * discourage_success: cat works, cat+ret fails, other op_success fails
 * verify_cat + discourage_success: everything fails

But that means that when adding "verify_cat" to "discourage_success" you're making valid uses of OP_CAT fail, which would be very confusing.

Reply:
@ajtowns Thanks for taking the time to elaborate in detail!

> If you wanted to have just two flags rather than three while still having each flag only add restrictions,

I think we are in agreement about this; that if you want to stick to the definition about flags only adding restrictions, then you need to do it the way you are suggesting.

My point was that is it worth sticking to this definition even when it makes the code more complex? 

I assume it is worth it, since you say it makes testing less confusing (and as you point out, the `discourage_cat` flag can be removed when the fork has been buried).

Reply:
> I think we are in agreement about this; that if you want to stick to the definition about flags only adding restrictions, then you need to do it the way you are suggesting.

It's consensus code -- you want to do everything you can to ensure that it works as intended and doesn't hide obscure bugs. It's also core policy; adding code in inquisition that breaks long-standing policy for the core repo is just going to make it hard to port the code across, which seems like it defeats the purpose.

> My point was that is it worth sticking to this definition even when it makes the code more complex?

I'm not sure I agree that it's more complex: it's an extra flag, and perhaps an extra line or two of code, but neither of those things are particularly complex. Likewise, I think that if we're planning on changing behaviour later, then implementing that behaviour by making it a dedicated flag, that we later treat as always-set or always-unset ("CAT has been active for years now, we will treat `DISCOURAGE_OP_CAT` as always unset") and remove entirely seems less complex than changing the behaviour of a flag we want to preserve ("`DISCOURAGE_OP_SUCCESS` will no longer affect `CAT` even when `VERIFY_TAPSCRIPT_OP_CAT` is unset").

But even if it were more complex to write in code, it's more important to have the design be simple and testable than for the code to be as simple as possible. Having the rule "each flag is purely restrictive" helps in a few ways ways: eg, it means that if we ensure that the mempool verification flags are a superset of the consensus flags, we'll never accept anything into the mempool that would be invalid in a block, it forces standardness features to be designed in a way that they could reasonably be upgraded to consensus features, and it gives a simple rule that we can (and do) automatically test.